include/pybind11/functional.h

```diff
@@ -38,14 +38,15 @@ struct type_caster<std::function<Return(Args...)>> {
 
         auto func = reinterpret_borrow<function>(src);
 
-        /*
-           When passing a C++ function as an argument to another C++
-           function via Python, every function call would normally involve
-           a full C++ -> Python -> C++ roundtrip, which can be prohibitive.
-           Here, we try to at least detect the case where the function is
-           stateless (i.e. function pointer or lambda function without
-           captured variables), in which case the roundtrip can be avoided.
-         */
+/*
+   When passing a C++ function as an argument to another C++
+   function via Python, every function call would normally involve
+   a full C++ -> Python -> C++ roundtrip, which can be prohibitive.
+   Here, we try to at least detect the case where the function is
+   stateless (i.e. function pointer or lambda function without
+   captured variables), in which case the roundtrip can be avoided.
+ */
+#ifndef BAD
         if (auto cfunc = func.cpp_function()) {
             auto *cfunc_self = PyCFunction_GET_SELF(cfunc.ptr());
             if (isinstance<capsule>(cfunc_self)) {
```

Comment:
@rwgk It looks like the issue is here where we have a capsule, but the contents of the capsule is not what we expect (a function record pointer). Therefore our reinterepret_cast is unsafe. We need a way to disambiguate function_record pointers from other contents in the capsule, thoughts @rwgk?

Reply:
General / without having looked very closely: AFAIK the capsule `name` is the only way to achieve some level of type safety.

Reply:
> we store it as nullptr.

The `name` is nullptr? (just to make sure I'm not misunderstanding)

Reply:
Yes.